src/main/java/org/jenkinsci/plugins/workflow/graph/StandardGraphLookupView.java

```diff
@@ -0,0 +1,237 @@
+ * Designed to work entirely on the basis of the {@link FlowNode#id} rather than the {@link FlowNode}s themselves.
+ */
+@SuppressFBWarnings(value = "ES_COMPARING_STRINGS_WITH_EQ", justification = "Can can use instance identity when comparing to a final constant")
+@Restricted(NoExternalUse.class)
+public final class StandardGraphLookupView implements GraphLookupView, GraphListener, GraphListener.Synchronous {
+
+    static final String INCOMPLETE = "";
+
+    /** Map the blockStartNode to its endNode, to accellerate a range of operations */
+    HashMap<String, String> blockStartToEnd = new HashMap<String, String>();
+
+    /** Map a node to its nearest enclosing block */
+    HashMap<String, String> nearestEnclosingBlock = new HashMap<String, String>();
+
+    /** Update with a new node added to the flowgraph */
+    public void onNewHead(@Nonnull FlowNode newHead) {
+        if (newHead instanceof BlockEndNode) {
+            blockStartToEnd.put(((BlockEndNode)newHead).getStartNode().getId(), newHead.getId());
+            String overallEnclosing = nearestEnclosingBlock.get(((BlockEndNode) newHead).getStartNode().getId());
+            if (overallEnclosing != null) {
+                nearestEnclosingBlock.put(newHead.getId(), overallEnclosing);
+            }
+        } else {
+            if (newHead instanceof BlockStartNode) {
+                blockStartToEnd.put(newHead.getId(), INCOMPLETE);
+            }
+
+            // Now we try to generate enclosing block info for caching, by looking at parents
+            // But we don't try to hard -- usually the cache is populated and we defer recursive walks of the graph
+            List<FlowNode> parents = newHead.getParents();
+            if (parents.size() > 0) {
+                FlowNode parent = parents.get(0);  // Multiple parents only for end of a parallel, and then both are BlockEndNodes
+
+                if (parent instanceof BlockStartNode) {
+                    nearestEnclosingBlock.put(newHead.getId(), parent.getId());
+                } else {
+                    // Try to reuse info from cache for this node:
+                    //   If the parent ended a block, we use info from the start of the block since that is at the same block nesting level as our new head
+                    //   Otherwise the parent is a normal atom node, and the head is enclosed by the same block
+                    String lookupId = (parent instanceof BlockEndNode) ? ((BlockEndNode) parent).getStartNode().getId() : parent.getId();
+                    String enclosingId = nearestEnclosingBlock.get(lookupId);
+                    if (enclosingId != null) {
+                        nearestEnclosingBlock.put(newHead.getId(), enclosingId);
+                    }
+                }
+            }
+        }
+    }
+
+    /** Create a lookup view for an execution */
+    public StandardGraphLookupView(FlowExecution execution) {
+        execution.addListener(this);
```

Comment:
It is an antipattern to attach an instance of some class to some external source (such as adding a listener) in its constructor when the class is not `final`—you can actually expose a partially-constructed object. Better to have the caller of the constructor do this.